src/sentry/region_to_control/producer.py

```diff
@@ -1,91 +1,178 @@
+)
+
+
+class AuditLogEntryService(InterfaceWithLifecycle):
+    @abstractmethod
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
+        pass
+
+
+class RegionToControlAuditLogEntryService(AuditLogEntryService):
+    region_to_control_messaging_service: RegionToControlMessagingService
+
+    def __init__(
+        self,
+        region_to_control_messaging_service: RegionToControlMessagingService = region_to_control_message_service,
+    ):
+        self.region_to_control_messaging_service = region_to_control_messaging_service
+
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
+        self.region_to_control_messaging_service.write_region_to_control_message(
             dataclasses.asdict(RegionToControlMessage(audit_log_event=entry.as_kafka_event())),
             sync=True,
         )
-    else:
+
+    def close(self):
+        pass
+
+
+class DatabaseBackedAuditLogEntryService(AuditLogEntryService):
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
         entry.save()
 
+    def close(self):
+        pass
 
-_publisher: Optional[KafkaProducer] = None
-_empty_headers = list()
 
+audit_log_entry_service: AuditLogEntryService = silo_mode_delegation(
+    {
+        SiloMode.MONOLITH: DatabaseBackedAuditLogEntryService,
+        SiloMode.CONTROL: DatabaseBackedAuditLogEntryService,
+        SiloMode.REGION: RegionToControlAuditLogEntryService,
+    }
+)
+
+
+class UserIpService(InterfaceWithLifecycle):
+    @abstractmethod
+    def produce_user_ip(self, event: UserIpEvent):
+        pass
+
+
+class RegionToControlUserIpService(UserIpService):
+    def close(self):
+        pass
+
+    region_to_control_messaging_service: RegionToControlMessagingService
 
-# TODO: Would be nice to configure the timeout via a dynamic setting.
-def write_region_to_control_message(payload: Any, sync=True, timeout=1):
-    future = get_region_to_control_producer().produce(
-        Topic(settings.KAFKA_REGION_TO_CONTROL),
-        KafkaPayload(
-            key=None,
-            value=json.dumps(payload).encode("utf8"),
-            headers=_empty_headers,
-        ),
-    )
-
-    if not sync:
-        return
-
-    future.result(timeout)
-
-
-def get_region_to_control_producer() -> KafkaProducer:
-    """
-    Creates, if necessary, an arroyo.KafkaProducer client configured for region to control communication and returns
-    it, caching it for future calls.  Installs an exit handler to close the worker thread processes.
-    """
-    global _publisher
-    if _publisher is None:
-        config = settings.KAFKA_TOPICS.get(settings.KAFKA_REGION_TO_CONTROL)
-        _publisher = KafkaProducer(
-            kafka_config.get_kafka_producer_cluster_options(config["cluster"])
+    def __init__(
+        self,
+        region_to_control_messaging_service: RegionToControlMessagingService = region_to_control_message_service,
+    ):
+        self.region_to_control_messaging_service = region_to_control_messaging_service
```

Comment:
Could call this `messaging_service` or `messenger` :shrug:

Reply:
🤷  I defer to others.  I'm super bad at making these kinds of decisions because I tend to read more code than I write lol.  But I think ti's a small enough surface to change at any time if helpful.